src/main/java/org/terasology/launcher/packages/PackageManager.java

```diff
@@ -109,6 +119,21 @@ public void initDatabase(Path launcherDir, Path gameDir) {
         );
     }
 
+    private void validateSchema(final Path jsonFile) {
+        try (
+                InputStream schemaIn = getClass().getResourceAsStream(SOURCES_SCHEMA);
+                InputStream jsonIn = Files.newInputStream(jsonFile)
+        ) {
+            final Schema schema = SchemaLoader.load(new JSONObject(new JSONTokener(schemaIn)));
+            schema.validate(new JSONArray((new JSONTokener(jsonIn))));
+        } catch (ValidationException e) {
+            logger.error("sources.json has invalid value at: {}", e.getPointerToViolation());
+            throw new RuntimeException("Failed to validate sources.json.");
```

Comment:
You don't need to throw a RuntimeException here.
Rather, make `validateScheme` return a `boolean` indicating whether or not it's valid.
Based on that information, we can decide how to continue. E.g. an alternative to a hard crash would be to show a user prompt on whether it's okay to replace the file with the default or suggest the player removes the `sources.json` and restarts the launcher.

Reply:
That'll be very user-friendly. I'm pushing a solution soon! :wink: